perfkitbenchmarker/background_tasks.py

```diff
@@ -0,0 +1,633 @@
+        receives worker_id when this thread's bootstrap code has completed and
+        receives a (worker_id, task_id) pair for each task completed on this
+        thread.
+  """
+  try:
+    response_queue.Put(worker_id)
+    while True:
+      task_tuple = task_queue.Get()
+      if task_tuple == _THREAD_STOP_PROCESSING:
+        break
+      elif task_tuple == _THREAD_WAIT_FOR_KEYBOARD_INTERRUPT:
+        while True:
+          time.sleep(_WAIT_MAX_RECHECK_DELAY)
+      task_id, task = task_tuple
+      task.Run()
+      response_queue.Put((worker_id, task_id))
+  except KeyboardInterrupt:
+    logging.exception('Child thread %s received a KeyboardInterrupt from its '
+                      'parent.', worker_id)
+
+
+class _BackgroundThreadTaskManager(_BackgroundTaskManager):
+  """Manages state for background tasks started in child threads."""
+
+  def __init__(self, *args, **kwargs):
+    super(_BackgroundThreadTaskManager, self).__init__(*args, **kwargs)
+    self._response_queue = _SingleReaderQueue()
+    self._task_queues = []
+    self._threads = []
+    self._available_worker_ids = range(self._max_concurrency)
+    uninitialized_worker_ids = set(self._available_worker_ids)
+    for worker_id in self._available_worker_ids:
+      task_queue = _NonPollingSingleReaderQueue()
+      self._task_queues.append(task_queue)
+      thread = threading.Thread(
+          target=_ExecuteBackgroundThreadTasks,
+          args=(worker_id, task_queue, self._response_queue))
+      thread.daemon = True
+      self._threads.append(thread)
+      thread.start()
+    # Wait for each Thread to finish its bootstrap code. Starting all the
+    # threads upfront like this and reusing them for later calls minimizes the
+    # risk of a KeyboardInterrupt interfering with any of the Lock interactions.
+    for _ in self._threads:
+      worker_id = self._response_queue.Get()
+      uninitialized_worker_ids.remove(worker_id)
+    assert not uninitialized_worker_ids, uninitialized_worker_ids
+
+  def __exit__(self, *unused_args, **unused_kwargs):
+    # Shut down worker threads.
+    for task_queue in self._task_queues:
+      task_queue.Put(_THREAD_STOP_PROCESSING)
+    for thread in self._threads:
+      _WaitForCondition(lambda: not thread.is_alive())
+
+  def StartTask(self, target, args, kwargs, thread_context):
+    assert self._available_worker_ids, ('StartTask called when no threads were '
+                                        'available')
+    task = _BackgroundTask(target, args, kwargs, thread_context)
+    task_id = len(self.tasks)
+    self.tasks.append(task)
+    worker_id = self._available_worker_ids.pop()
+    self._task_queues[worker_id].Put((task_id, task))
+
+  def AwaitAnyTask(self):
+    worker_id, task_id = self._response_queue.Get()
+    self._available_worker_ids.append(worker_id)
+    return task_id
+
+  def HandleKeyboardInterrupt(self):
+    # Raise a KeyboardInterrupt in each child thread.
+    for thread in self._threads:
```

Comment:
It could be helpful to know where they were interrupted. But at the same time, threads that are not actively processing a task will likely be interrupted while within _ExecuteBackgroundThreadTasks, so that could be a lot of useless information. The only way that I can think of to differentiate helpful from unhelpful cases is to inspect the stack of the exception, which I would want to be a separate thread. How about if I switch this to debug level for now (but still include the exc_info) and make a separate PR to reduce output?